### PR TITLE
AMDGPU: Add v_permlane16_swap_b32 and v_permlane32_swap_b32 for gfx950

### PR DESCRIPTION
**Pull request by arsenm**
This was a bit annoying because these introduce a new special case
encoding usage. op_sel is repurposed as a subset of dpp controls,
and is eligible for VOP3->VOP1 shrinking. For some reason fi also
uses an enum value, so we need to convert the raw boolean to 1 instead
of -1.

The 2 registers are swapped, so this has 2 defs. Ideally the builtin
would return a pair, but that's difficult so return a vector instead.
This would make a hypothetical builtin that supports v2f16 directly
uglier.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117260?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#117261** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117261?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117260** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117260?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117260?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#117259** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117259?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117258** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117258?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117257** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117257?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117256** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117256?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117235** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117235?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117234** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117233** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117233?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117232** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117232?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117214** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117214?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117213** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117213?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117212** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117211** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117205** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117202** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117055** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117055?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117053** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117053?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117052** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117052?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116728** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116728?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116724** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116724?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#117047](https://github.com/llvm/llvm-project/pull/117047) <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117047?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#116723** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116723?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116722** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116722?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116681** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116681?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116680** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116680?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116679** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116678** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116678?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116312** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116311** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116310** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116310?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116308** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116307** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116307?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: llvm/lib/Target/AMDGPU/GCNSubtarget.h [hunk 1]
```diff
@@ -1319,6 +1321,10 @@ class GCNSubtarget final : public AMDGPUGenSubtargetInfo,
   /// \returns true if the target has instructions with xf32 format support.
   bool hasXF32Insts() const { return HasXF32Insts; }
 
+  bool hasPrngInst() const { return HasPrngInst; }
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/GCNSubtarget.h:1324]: This change is unrelated.

**Review by shiltian — APPROVED** [review]

**Comment by arsenm** [comment]
### Merge activity

* **Nov 22, 11:08 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/117260).
